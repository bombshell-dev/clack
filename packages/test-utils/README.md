# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
