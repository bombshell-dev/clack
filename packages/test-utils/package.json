{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
