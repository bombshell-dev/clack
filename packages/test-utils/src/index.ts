export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
