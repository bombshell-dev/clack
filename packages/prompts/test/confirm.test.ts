import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest';
import * as prompts from '../src/index.js';
import { MockReadable, MockWritable } from './test-utils.js';

describe.each(['true', 'false'])('confirm (isCI = %s)', (isCI) => {
	let originalCI: string | undefined;
	let output: MockWritable;
	let input: MockReadable;

	beforeAll(() => {
		originalCI = process.env.CI;
		process.env.CI = isCI;
	});

	afterAll(() => {
		process.env.CI = originalCI;
	});

	beforeEach(() => {
		output = new MockWritable();
		input = new MockReadable();
	});

	afterEach(() => {
		vi.restoreAllMocks();
	});

	test('renders message with choices', async () => {
		const result = prompts.confirm({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders custom active choice', async () => {
		const result = prompts.confirm({
			message: 'foo',
			active: 'bleep',
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders custom inactive choice', async () => {
		const result = prompts.confirm({
			message: 'foo',
			inactive: 'bleep',
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});

	test('right arrow moves to next choice', async () => {
		const result = prompts.confirm({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', 'right', { name: 'right' });
		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe(false);
		expect(output.buffer).toMatchSnapshot();
	});

	test('left arrow moves to previous choice', async () => {
		const result = prompts.confirm({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', 'right', { name: 'right' });
		input.emit('keypress', 'left', { name: 'left' });
		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});

	test('can cancel', async () => {
		const result = prompts.confirm({
			message: 'foo',
			input,
			output,
		});

		input.emit('keypress', 'escape', { name: 'escape' });

		const value = await result;

		expect(prompts.isCancel(value)).toBe(true);
		expect(output.buffer).toMatchSnapshot();
	});

	test('can set initialValue', async () => {
		const result = prompts.confirm({
			message: 'foo',
			initialValue: false,
			input,
			output,
		});

		input.emit('keypress', '', { name: 'return' });

		const value = await result;

		expect(value).toBe(false);
		expect(output.buffer).toMatchSnapshot();
	});
});
