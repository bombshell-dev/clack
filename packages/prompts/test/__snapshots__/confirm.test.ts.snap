// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`confirm (isCI = false) > can cancel 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[31m■[39m  foo
[90m│[39m  [9m[2mNo[22m[29m
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = false) > can set initialValue 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [2m○[22m [2mYes[22m [2m/[22m [32m●[39m No
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mNo[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = false) > left arrow moves to previous choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  [2m○[22m [2mYes[22m [2m/[22m [32m●[39m No",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = false) > renders custom active choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m bleep [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mbleep[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = false) > renders custom inactive choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mbleep[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = false) > renders message with choices 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = false) > right arrow moves to next choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  [2m○[22m [2mYes[22m [2m/[22m [32m●[39m No",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mNo[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > can cancel 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[31m■[39m  foo
[90m│[39m  [9m[2mNo[22m[29m
[90m│[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > can set initialValue 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [2m○[22m [2mYes[22m [2m/[22m [32m●[39m No
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mNo[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > left arrow moves to previous choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  [2m○[22m [2mYes[22m [2m/[22m [32m●[39m No",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > renders custom active choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m bleep [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mbleep[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > renders custom inactive choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mbleep[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > renders message with choices 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mYes[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`confirm (isCI = true) > right arrow moves to next choice 1`] = `
[
  "<cursor.hide>",
  "[90m│[39m
[36m◆[39m  foo
[36m│[39m  [32m●[39m Yes [2m/[22m [2m○[22m [2mNo[22m
[36m└[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36m│[39m  [2m○[22m [2mYes[22m [2m/[22m [32m●[39m No",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32m◇[39m  foo
[90m│[39m  [2mNo[22m",
  "
",
  "<cursor.show>",
]
`;
